### ISSUE TITLE
vulnerabilities severity shows not the correct severity

### ISSUE THREAD

**Issue by akitzing** [goharbor/harbor#14929]
In the artefacts overview of a project it shows for example critical but the highest severity is in the detailed view is only high.
used version 2.2.1 with trivy.

**Comment by danielpacak** [comment]
I checked https://github.com/aquasecurity/harbor-scanner-trivy and the code for calculating overall severity looks okey. I also tried a few images and I could not reproduce this issue in my environment with Harbor v2.2.1 deployed with the installer script.

@akitzing Did you change by any chance the default Trivy configuration in `harbor.yaml`? There are some settings to skip unfixed vulnerabilities or choose severities. Thinking out loud this may cause some trouble even though it's very unlikely.

It would be very useful if you could share a minimal reproducible container image so we can load it into our test env and compare results.

**Comment by akitzing** [comment]
the harbor.yml has no subsequent changes. (It is the configuration from the previous versions)
I took the newest 
![Harbor-Scan](https://user-images.githubusercontent.com/63971609/123923526-626d2880-d989-11eb-8dc9-bfca110f6ae5.jpg)
image grafana/promtail:latest. A critical is displayed here. 21 High, 8 Medium, 73 Low ...

**Comment by danielpacak** [comment]
Thanks for sharing @akitzing but still I cannot reproduce in my environment as shown below.

1. Make sure that Trivy adapter is using the latest vulnerabilities database
   ![trivy_adapter_and_config](https://user-images.githubusercontent.com/1322923/124037118-5754df80-d9ff-11eb-9906-7364aa2c141c.png)
2. Confirm that the summary show 2 critical vulnerabilities
   ![summary_de](https://user-images.githubusercontent.com/1322923/124037161-689dec00-d9ff-11eb-89ac-1d0bd6fa125a.png)
3. Confirm that the full listing also shows 2 critical vulnerabilities
   ![list_vulnerabilities](https://user-images.githubusercontent.com/1322923/124037405-cd594680-d9ff-11eb-97d3-55ffb8e68ff7.png)

Maybe this is only reproducible in case of multiple artefacts in the same repository. @akitzing could you share the image digest of the artifact that starts with the `sha256:bf68acc2` digest?

**Comment by akitzing** [comment]
1.  I have exactly the same database values.
2. I have no critical vulnerabilities in the overview. The two critical ones CVE-2021-33574 are unknowns. 
3. (Here it looks better)
```
scanner [ / ]$ trivy image grafana/promtail:latest
2021-07-01T06:40:49.852Z        WARN    You should avoid using the :latest tag as it is cached. You need to specify '--clear-cache' option when :latest image is changed
2021-07-01T06:40:49.874Z        INFO    Need to update DB
2021-07-01T06:40:49.875Z        INFO    Downloading DB...
22.26 MiB / 22.26 MiB [-----------------------------------------------------------] 100.00% 8.64 MiB p/s 2s
2021-07-01T06:40:55.163Z        INFO    Detecting Debian vulnerabilities...
2021-07-01T06:40:55.177Z        INFO    Trivy skips scanning programming language libraries because no supported file was detected

grafana/promtail:latest (debian 10.10)
======================================
Total: 106 (UNKNOWN: 2, LOW: 75, MEDIUM: 8, HIGH: 19, CRITICAL: 2)
```

grafana/promtail@sha256:bf68acc249098107347e736c8be897a80f97fbac0a85beed86f41d0d53546864

**Comment by danielpacak** [comment]
Okey, at least we have consistent results when we run trivy command 😅 However, even with two artifacts pushed to the repository I still cannot reproduce ...

So we need to narrow down whether the issue is in Harbor UI (JavaScript or Web browser caching), Harbor API or Harbor Database (bug in code or data consistency).

Let's start with an infamous ask. 🤓 Could you empty caches in your Web browser and refresh the page and see if that helps. If not continue reading.

Inspect the JSON output returned by Harbor API server by looking for such XHR request
http://harbor.domain/api/v2.0/projects/grafana/repositories/promtail/artifacts/sha256:c8688317ebcc2effdf15eb5a4252d1907235aa1e24940538397fa39cd4af398b/additions/vulnerabilities. Confirm the value of the `severity` property.

![vulnerabilities_xhr](https://user-images.githubusercontent.com/1322923/124090582-420aa000-da55-11eb-8ee1-7a51ecbb81fc.png)

The last thing to try out is to check the data in the database, but I'll wait for you confirmation what is the value returned by Harbor API as JSON.

**Comment by akitzing** [comment]
No success with the browser.
I also ran the API query from another machine. It remains with the Critical.  
In the whole output there is only one Critical. No critical vulnerabilities.

`{"application/vnd.scanner.adapter.vuln.report.harbor+json; version=1.0":{"generated_at":"2021-07-01T10:06:55.620255746Z"                                          ,"scanner":{"name":"Trivy","vendor":"Aqua Security","version":"v0.16.0"},"severity":"Critical","vulnerabilities":[{"id":                                          "CVE-2019-18276","package":"bash","version":"5.0-4","fix_version":`

**Comment by danielpacak** [comment]
To confirm that I understood so you're saying that there overall severity in critical (`severity` property) whereas there are no critical vulnerabilities in the list returned by the API (`vulnerabilities` property).

If that's the case we must pick into the database and resolve this mystery 💪 

Could you run the following queries in your environment and share the results? Essentially this is how we query the. database to build the list of vulnerabilities and the summary.

```sql
-- Query vulnerabilities for a specified artifact digest
-- That's what Harbor UI displays in vulnerabilities listing
SELECT vr.cve_id, severity, cvss_score_v3, package,package_version, fixed_version FROM scan_report sr
  JOIN report_vulnerability_record rvr on sr.uuid = rvr.report_uuid
  JOIN vulnerability_record vr on rvr.vuln_record_id = vr.id
 WHERE sr.digest = 'sha256:c8688317ebcc2effdf15eb5a4252d1907235aa1e24940538397fa39cd4af398b'
   AND sr.mime_type = 'application/vnd.scanner.adapter.vuln.report.harbor+json; version=1.0'
   AND vr.cve_id in ('CVE-2021-33574');
```

| cve\_id | severity | cvss\_score\_v3 | package | package\_version | fixed\_version |
| :--- | :--- | :--- | :--- | :--- | :--- |
| CVE-2021-33574 | Critical | 9.8 | libc-bin | 2.28-10 |  |
| CVE-2021-33574 | Critical | 9.8 | libc6 | 2.28-10 |  |

```sql
-- Query the summary of vulnerabilities count grouped by severity
-- That's what Harbor UI displays in the summary tooltip
SELECT vr.severity, count(vr.severity) FROM scan_report sr
  JOIN report_vulnerability_record rvr on sr.uuid = rvr.report_uuid
  JOIN vulnerability_record vr on rvr.vuln_record_id = vr.id
 WHERE sr.digest = 'sha256:c8688317ebcc2effdf15eb5a4252d1907235aa1e24940538397fa39cd4af398b'
   AND sr.mime_type = 'application/vnd.scanner.adapter.vuln.report.harbor+json; version=1.0'
 GROUP BY vr.severity;
```

| severity | count |
| :--- | :--- |
| Medium | 8 |
| Critical | 2 |
| Unknown | 2 |
| Low | 75 |
| High | 19 |

**Comment by akitzing** [comment]
```
     cve_id     | severity | cvss_score_v3 | package  | package_version | fixed_version 
----------------+----------+---------------+----------+-----------------+---------------
 CVE-2021-33574 | Unknown  |               | libc-bin | 2.28-10         | 
 CVE-2021-33574 | Unknown  |               | libc6    | 2.28-10         | 
(2 rows)
```

```
severity | count 
----------+-------
 Medium   |     8
 Unknown  |     4
 Low      |    73
 High     |    21
```

**Comment by danielpacak** [comment]
It seems that you have outdated information in the database. I tried to reproduce by manually setting severity of CVE-2021-33574 to Unknown and rescanING the image, which should overwrite the severity back to Critical, but it's not the case. Looks to me like a bug in Harbor when we create or update records in the `vulnerability_record` table. /cc @prahaladdarkin 

To further investigate why we ended up in such data inconsistency, could you provide:
1. Details about the DB. Is it external DB or the default one that comes with the Harbor installer?
2. Is is a fresh install of Harbor or you used migration scripts. If migration please specify versions.
3. Check that there are no errors in Harbor core related to SQL and DB?

A quick check to see if the data consistency is the problem in your environment is to delete CVE-2021-33574 manually from the database and rescan. Notice however that this is **potentially dangerous operation** that might impact other vulnerability reports stored in the DB. Make sure that you have a **DB backup or time and capacity to rescan** all artifacts in the registry.

```sql
DELETE FROM vulnerability_record
WHERE cve_id = 'CVE-2021-33574'
```

**Comment by akitzing** [comment]
1. it comes with the installer
2. it was migrated with adjustments to docker-compose.yml. I think it was from version 2.0.x -> 2.1 -> 2.2.1
3. no errors related to database
4. Clair was used in previous versions but is removed

Planned to migrate to version 2.3, maybe this will help?

**Comment by danielpacak** [comment]
Actually I looked it up and in the [v2.2.0 release notes](https://github.com/goharbor/harbor/releases/tag/v2.2.0) we mention that existing vulnerability reports must be deleted. There was non-backward compatible change in code and new normalized DB schema to persist vulnerability data. See **Breaking Changes** section.

**Comment by akitzing** [comment]
After deleting the table and rescanned it, the correct status for the "promtail image" was displayed.
I went through many images, but on another image I noticed that it is now displayed the other way around. Severity: High, but it has critical vulnerabilities.
onlyoffice/documentserver:6.3.0 | sha256:1dfb7668fa18649cf62edb1c906a699b489ecabd0fdda9de7e5a7e399ffeb6c1

```
registry=# SELECT vr.severity, count(vr.severity) FROM scan_report sr
registry-#   JOIN report_vulnerability_record rvr ON sr.uuid = rvr.report_uuid
registry-#   JOIN vulnerability_record vr ON rvr.vuln_record_id = vr.id
registry-#  WHERE sr.digest = 'sha256:1dfb7668fa18649cf62edb1c906a699b489ecabd0fdda9de7e5a7e399ffeb6c1'
registry-#    AND sr.mime_type = 'application/vnd.scanner.adapter.vuln.report.harbor+json; version=1.0'
registry-#  GROUP BY vr.severity;
 severity | count 
----------+-------
 Medium   |   132
 Critical |    10
 Low      |   121
 High     |     5
(4 rows)
```


![image](https://user-images.githubusercontent.com/63971609/124247910-a182b180-db22-11eb-9320-21f68818322d.png)

`{"application/vnd.scanner.adapter.vuln.report.harbor+json; version=1.0":{"generated_at":"2021-07-02T08:42:00.665776178Z","scanner":{"name":"Trivy","vendor":"Aqua Security","version":"v0.16.0"},"severity":"High","vulnerabilities":...`

**Comment by heww** [comment]
This issue may duplicate with #14745 which was fixed in 2.3.x

**Comment by akitzing** [comment]
I updated and it works.
